### DISCUSSION TITLE
External lib without type hints, override method return type

### DISCUSSION THREAD

**Discussion by choucavalier** [microsoft/pyright#6736] [General]
I'm using a class `Foo` from an external library that does not have any type hints. This class contains a default implementation that returns an empty list. In reality, this library can return anything from an instance of another class `Bar`, a list of instances of `Bar`, or a dictionary mapping a `str` to a `Bar`. In other words, the actual return type is `Bar | list[Bar] | dict[str, Bar]`. But this type hint is not provided by the external library.

```python
class Foo:
  def some_method(self):
    return [] # default implementation
```

In my code, I want to override the return type of this method, to make it clear what it can be.

```python
class MyClass(the_lib.Foo):
  some_method: Callable[[], Bar | list[Bar] | dict[str, Bar]]
```

Sadly, pyright seems to prefer the inferred type of `Foo` in that library over my own type annotations. It results in an `define method "some_method" in an incompatible way`, and more precisely `Return type mismatch: base method returns type "Bar | list[Bar] | dict[str, Bar]", override returns type "list[Unknown]"`.

What is the right way to force pyright to use my type annotation rather than the inferred type based on the library I have no control over?

**Comment by erictraut** [comment]
Pyright will use your type annotation if you are using an instance of `MyClass`, but you've overridden the method in an incorrect manner. A variable that is a `Callable` is not the same as an instance method. In particular, your `Callable` doesn't accept a `self` parameter.

Try this instead...

```python
from typing import Callable

class Foo:
    def some_method(self):
        return []  # default implementation

class Bar:
    ...

class MyClass(Foo):
    def some_method(self) -> Bar | list[Bar] | dict[str, Bar]:
        return super().some_method()
```

**Reply by choucavalier** [reply]
Hi Eric, thanks for taking the time. This is exactly what I first tried and I got an `reportIncompatibleMethodOverride` error.

```
import luigi

from my_lib import Bar


class BaseClass(luigi.Task):
    def output(self) -> Bar | dict[str, Bar] | list[Bar]:
        return {}
```

```
Method "output" overrides class "Task" in an incompatible manner Pyright (reportIncompatibleMethodOverride) [10, 9]                         
  Return type mismatch: base method returns type "list[Unknown]", override returns type "Bar | dict[str, Bar] | list[Bar]"
    Type "Bar | dict[str, Bar] | list[Bar]" cannot be assigned to type "list[Unknown]"                                    
      "Bar" is incompatible with "list[Unknown]"                                                                                    
```

**Reply by erictraut** [reply]
Yes, that's correct. This violates the `reportIncompatibleMethodOverride` type rule, so pyright is correct to log an error here. Starting with pyright 1.1.339, `reportIncompatibleMethodOverride` is enabled by default because it's required for compliance with the Python static typing spec. You can disable this diagnostic rule in your configuration if you'd prefer.

If you want to keep this diagnostic rule enabled, you have a couple of options:
1. Install (or create, if necessary) a type stub for the library that properly defines the return type.
2. Suppress this diagnostic on a per-file basis using `# pyright: reportIncompatibleMethodOverride=false` at the top of the file.
3. Suppress this diagnostic on this like by using a `# pyright: ignore[reportIncompatibleMethodOverride]` end-of-line comment.

**Reply by choucavalier** [reply]
Thanks. Is there a way to ignore the report in the `BaseClass` and make all subclasses of `BaseClass` assume the right signature to be the one defined in `BaseClass` rather than the one defined in the imported library's definition (`luigi.Task`)?

**Reply by erictraut** [reply]
You can create a type stub for the library. Or you can work with the maintainer of the library to get the proper type annotations in place.

**Reply by choucavalier** [reply]
Ok, I see. Thanks again Eric for taking some of your precious time to answer the community. This is very precious :)

**Reply by Viicos** [reply]
One hacky workaround, if you can't/don't want to apply the solutions provided (or if you are waiting on proper types to be merged in the upstream library):

```python
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    class Foo:
        ...  # Define the typed method you are looking for
else:
    from lib import Foo
```

Although you might have to copy the whole `Foo` class definition in the `if TYPE_CHECKING` block

**Comment by choucavalier** [comment]
I checked again with the latest pyright version `1.1.340`

```python
class Mama:
    def output(self):
        return []

class BaseTask(Mama):
    def output(self) -> list[str] | str | dict[str, str]:
        raise NotImplementedError()
```

this is still yielding a `reportIncompatibleMethodOverride`
